modules/auxiliary/cloud/aws/enum_ssm.rb

```diff
@@ -0,0 +1,153 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'aws-sdk-ssm'
+
+class MetasploitModule < Msf::Auxiliary
+  include Rex::Proto::Http::WebSocket::AmazonSsm
+  include Msf::Auxiliary::CommandShell
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'        => 'Amazon Web Services EC2 instance enumeration',
+        'Description' => %q(
+                          Provided AWS credentials, this module will call the authenticated
+                          API of Amazon Web Services to list all SSM-enabled EC2 instances
+                          accessible to the account. Once enumerated as SSM-enabled, the
+                          instances can be controlled using out-of-band WebSocket sessions
+                          provided by the AWS API (nominally, privileged out of the box).
+                          This module provides not only the API enumeration identifying EC2
+                          instances accessible via SSM with given credentials, but enables
+                          session initiation for all identified targets (without requiring
+                          target-level credentials). The module also provides an EC2 ID
+                          filter and a limiting throttle to prevent session stampedes or
+                          expensive messes.
+                         ),
+        'Author'      => [
+          'Aaron Soto <aaron.soto@rapid7.com>', # EC2 enum module
+          'RageLtMan <rageltman[at]sempervictus>' # SSM stuff
+        ],
+        'License'     => MSF_LICENSE
+      )
+    )
+
+    register_options(
+      [
+        OptBool.new('GET_SSM_SESSION', [true, 'Automatically get SSM sessions once found', false]),
+        OptInt.new('LIMIT', [false, 'Only return the specified number of results from each region']),
+        OptString.new('FILTER_EC2_ID', [false, 'Look for specific EC2 instance ID']),
+        OptString.new('REGION', [false, 'AWS Region (eg. "us-west-2")']),
+        OptString.new('ACCESS_KEY_ID', [true, 'AWS Access Key ID (eg. "AKIAXXXXXXXXXXXXXXXX")', '']),
+        OptString.new('SECRET_ACCESS_KEY', [true, 'AWS Secret Access Key (eg. "CA1+XXXXXXXXXXXXXXXXXXXXXX6aYDHHCBuLuV79")', ''])
+      ]
+    )
+  end
+
+  def handle_aws_errors(e)
+    if e.class.module_parents.include?(Aws)
+      fail_with(Failure::UnexpectedReply, e.message)
+    else
+      raise e
+    end
+  end
+
+  def enumerate_regions
+    regions = []
+
+    ec2 = Aws::EC2::Resource.new(
```

Comment:
Getting this error here. Also shouldn't the region be `datastore['REGION']`?

```
msf6 auxiliary(cloud/aws/enum_ssm) > run

[-] Auxiliary failed: NameError uninitialized constant Aws::EC2
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:60:in `enumerate_regions'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:75:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(cloud/aws/enum_ssm) >
```

Reply:
Pretty sure upstream has the `ec2` SDK... i can add it to the `require` for the module, but was under impression that Zeitwerk handles all of that. I'm still not 100% with how all the modern dependency management stuff works.

Reply:
Still getting a stack trace here when REGION is not specified.

```
msf6 auxiliary(cloud/aws/enum_ssm) > unset REGION
Unsetting REGION...
msf6 auxiliary(cloud/aws/enum_ssm) > run

[-] Auxiliary failed: Aws::Errors::MissingRegionError No region was provided. Configure the `:region` option or export the region name to ENV['AWS_REGION']
[-] Call stack:
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/aws-sdk-core/plugins/regional_endpoint.rb:92:in `after_initialize'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/seahorse/client/base.rb:81:in `block in after_initialize'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/seahorse/client/base.rb:80:in `each'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/seahorse/client/base.rb:80:in `after_initialize'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/seahorse/client/base.rb:24:in `initialize'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-ec2-1.354.0/lib/aws-sdk-ec2/client.rb:370:in `initialize'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-core-3.168.3/lib/seahorse/client/base.rb:102:in `new'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.4@metasploit-framework/gems/aws-sdk-ec2-1.354.0/lib/aws-sdk-ec2/resource.rb:28:in `initialize'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:61:in `new'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:61:in `enumerate_regions'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:76:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(cloud/aws/enum_ssm) > 
```